ssl/statem/statem_clnt.c

```diff
@@ -821,7 +821,10 @@ int tls_construct_client_hello(SSL *s)
         goto err;
     }
 
-    if ((sess == NULL) || (sess->ssl_version != s->version) ||
+    if ((sess == NULL) ||
+        /* Only destroy session which version is bigger than maximum supported.
+         */
+        ssl_version_cmp(s, sess->ssl_version, s->version) > 0 ||
```

Comment:
It's more "doomed to be rejected" rather than "doomed to fail" (i.e. it's not a problem for the connection, just pointless to offer), but that's a good point. The check should actually be the same as `tls_process_server_hello` would do, but without the actual version switch.

(I should go fix that in BoringSSL too. Thanks!)